tests/test_spark.py

```diff
@@ -0,0 +1,123 @@
+from granulate_utils.metrics.metrics import (
+    SPARK_AGGREGATED_STAGE_METRICS,
+    SPARK_APPLICATION_DIFF_METRICS,
+    SPARK_APPLICATION_GAUGE_METRICS,
+    SPARK_EXECUTORS_METRICS,
+)
+from granulate_utils.metrics.sampler import BigDataSampler
+from pytest import LogCaptureFixture
+
+from gprofiler.log import get_logger_adapter
+from tests.conftest import _build_image
+from tests.utils import _wait_container_to_start
+
+# `BigDataSampler` receives a logger as an argument.
+logger = get_logger_adapter("gprofiler_spark_test")
+
+# List that includes all the metrics that are expected to be collected by `BigDataSampler` in SA mode.
+EXPECTED_SA_METRICS_KEYS = [
+    metric
+    for metrics_dict in (
+        SPARK_APPLICATION_GAUGE_METRICS,
+        SPARK_APPLICATION_DIFF_METRICS,
+        SPARK_AGGREGATED_STAGE_METRICS,
+        SPARK_EXECUTORS_METRICS,
+    )
+    for metric in metrics_dict.values()
+]
+
+SPARK_MASTER_HOST = "127.0.0.1"
+
+DISCOVER_INTERVAL_SECS = 5
+DISCOVER_TIMEOUT_SECS = 60
+
+
+@pytest.fixture(scope="function")
+def sparkpi_container(docker_client: DockerClient, application_docker_mounts: List[Mount]) -> Container:
+    """
+    This fixture is responsible for running SparkPi application in a container.
+    See `containers/spark/Dockerfile`
+    """
+    spark_image = _build_image(docker_client=docker_client, runtime="spark")
+    container = docker_client.containers.run(
+        spark_image,
+        detach=True,
+        mounts=application_docker_mounts,
+        network_mode="host",
+        pid_mode="host",
+        environment={"SPARK_MASTER_HOST": SPARK_MASTER_HOST},
+    )
+    _wait_container_to_start(container)
+    try:
+        yield container
+    finally:
+        container.stop()
+        container.remove()
+
+
+def _validate_sa_metricssnapshot(snapshot: MetricsSnapshot) -> None:
+    """
+    Validates that the snapshot contains all the expected metrics.
+    """
+    samples = snapshot.samples
+    assert len(samples) != 0, "No samples found in snapshot"
+    metric_keys = [sample.name for sample in samples]
+    for key in EXPECTED_SA_METRICS_KEYS:
+        assert key in metric_keys, f"Metric {key} not found in snapshot"
+
+
+def test_sa_spark_discovered_mode(caplog: LogCaptureFixture, sparkpi_container: Container) -> None:
+    """
+    Validates `BigDataSampler`s' `discover()` and `snapshot()` API's in discover mode.
+    In discover mode we do not know what's the cluster mode and master address.
+    """
+    discover_timeout = DISCOVER_TIMEOUT_SECS
+    caplog.set_level(logging.DEBUG)
+    sampler = BigDataSampler(logger, "", None, None, False)
+    # We want to make the discovery process as close as possible to the real world scenario.
+    while not (discovered := sampler.discover()) and discover_timeout > 0:
+        sleep(DISCOVER_INTERVAL_SECS)
+        discover_timeout -= DISCOVER_INTERVAL_SECS
+    assert discovered, "Failed to discover cluster mode and master address"
```

Comment:
Add assert on the actual expected discovered cluster settings

Reply:
03972fb